rfcs/20201221-tfmot-compression-api.md

```diff
@@ -0,0 +1,447 @@
+
+During the training phase, `project_training_weights` method is called for each training step. After fine-tuning training for compression is finished, we convert the training phase model to a compressed model. We only call the `compress_training_weights` function once for each compressible tensor for converting.
+
+Compressed model contains the `decompress_weights` function in the graph. It’s possible to call the `decompress_weights` for each inference step. To improve performance, we’ll cache the decompressed one depending on flags if we have enough space.
+
+```python
+class WeightCompressionAlgorithm(metaclass=abc.ABCMeta):
+  """Interface for weight compression algorithm that acts on a per-layer basis.
+
+     This allows both options of either decompressing during inference or
+     decompressing prior to inference (where compression occurs by applying a
+     tool such as zip to the model file).
+
+     This interface is a purely functional one.
+  """
+
+  @abc.abstractmethod
+  def get_compressible_weights(
+      self, original_layer: tf.keras.layers.Layer) -> List[str]:
+    """Define compressible weights for each layer.
+
+    Args:
+       original_layer: tf.keras.layers.Layer representing a layer from the
+       original model.
+
+    Returns:
+       List of attribute names as string representing list of compressible
+       weights for the given layer. (e.g. return value ['kernel'] means
+       layer.kernel is compressible.)
+    """
+
+  @abc.abstractmethod
+  def init_training_weights(
+      self, pretrained_weight: tf.Tensor):
+    """Initialize training weights for the training model. It calls the `add_training_weight` method several times to add training weights.
+
+    Args:
+      pretrained_weight: tf.Tensor of a pretrained weight of a layer that will
+        be compressed eventually.
+    """
+
+  def add_training_weight(
+      self, *args, **kwargs):
+    """Add training weight for the training model. This method is called from `init_training_weights`.
+
+    Args:
+      *args, **kwargs: args and kwargs for training_model.add_weight.
+    """
+
+  @abc.abstractmethod
+  def project_training_weights(self, *training_weights: tf.Tensor) -> tf.Tensor:
+    """Define a piece of the forward pass during training, which operates on a single compressible weight.
+    The default throws an error when training occurs.
+
+    Args:
+       *training_weights: tf.Tensors representing any variables used during
+         training, for a single compressible weight, in the order returned in
+         `init_training_weights`.
+
+    Returns:
+       tf.Tensor to set the compressible weight to.
+    """
+
+  def update_training_weight(self, index: integer, tensor: tf.Tensor):
```

Comment:
It seems quite error-prone to refer to a weight variable by its index. Prefer passing the actual weight variable.

Reply:
That make sense. Let's change this to tf.Tensor instead of integer.